File: slp/backend-dotnet/Features/Explanation/ExplanationRepository.cs
using Microsoft.EntityFrameworkCore;
using backend_dotnet.Data;

namespace backend_dotnet.Features.Explanation;

public class ExplanationRepository : IExplanationRepository
{
    private readonly AppDbContext _context;

    public ExplanationRepository(AppDbContext context)
    {
        _context = context;
    }

    /// <summary>
    /// Returns explanations for a source visible to this user:
    /// system explanations (user_id IS NULL) + this user's own explanations.
    /// </summary>
    public async Task<IEnumerable<Explanation>> GetBySourceIdAsync(int sourceId, int userId)
    {
        return await _context.Explanations
            .Where(e => e.SourceId == sourceId &&
                        (e.UserId == null || e.UserId == userId))
            .OrderBy(e => e.CreatedAt)
            .ToListAsync();
    }

    public async Task<Explanation?> GetByIdAsync(int id)
    {
        return await _context.Explanations.FindAsync(id);
    }

    public async Task<Explanation> CreateAsync(Explanation explanation)
    {
        _context.Explanations.Add(explanation);
        await _context.SaveChangesAsync();
        return explanation;
    }

    public async Task UpdateAsync(Explanation explanation)
    {
        explanation.UpdatedAt = DateTime.UtcNow;
        _context.Explanations.Update(explanation);
        await _context.SaveChangesAsync();
    }

    public async Task DeleteAsync(int id)
    {
        var entity = await _context.Explanations.FindAsync(id);
        if (entity != null)
        {
            _context.Explanations.Remove(entity);
            await _context.SaveChangesAsync();
        }
    }
}

File: slp/backend-nestjs/src/modules/explanation/explanation.repository.ts
import { Injectable } from '@nestjs/common';
import { DataSource, Repository } from 'typeorm';
import { Explanation } from './explanation.entity';

@Injectable()
export class ExplanationRepository {
  private repo: Repository<Explanation>;

  constructor(private dataSource: DataSource) {
    this.repo = this.dataSource.getRepository(Explanation);
  }

  /**
   * Returns explanations for a source visible to the user:
   * system explanations (user_id IS NULL) + this user's own explanations.
   */
  async getBySourceId(sourceId: number, userId: number): Promise<Explanation[]> {
    return this.repo
      .createQueryBuilder('e')
      .where('e.sourceId = :sourceId', { sourceId })
      .andWhere('(e.userId IS NULL OR e.userId = :userId)', { userId })
      .orderBy('e.createdAt', 'ASC')
      .getMany();
  }

  async getById(id: number): Promise<Explanation | null> {
    return this.repo.findOne({ where: { id } });
  }

  async create(entity: Explanation): Promise<Explanation> {
    const saved = await this.repo.save(entity);
    return saved;
  }

  async update(entity: Explanation): Promise<Explanation> {
    entity.updatedAt = new Date();
    return this.repo.save(entity);
  }

  async delete(id: number): Promise<void> {
    await this.repo.delete(id);
  }
}